### ISSUE TITLE
Pac-Man Museum Plus

### ISSUE THREAD

**Issue by bestgithubaccount1** [farmerbb/RED-Project#98]
![image](https://user-images.githubusercontent.com/63813081/231871684-8340ccc2-15e5-4013-b828-092296420d28.png)
It appears that Namco used the same format for Pac-Man Museum Plus as they did for Arcade Game Series for compressing the emulated files. Any chance these could be extracted? (Also has any progress been made on Ms. Pac-Man Arcade Game Series?)

**Comment by Computerlabrat** [comment]
Ms Pac-Man? Maybe [this](https://pypi.org/project/game-extraction-toolbox/) will help.

**Comment by kjetil-f** [comment]
Seems possible with Game Extraction Toolbox #154

**Comment by ghost** [comment]
Thought I'd throw my hat in the ring and share some info on a few of the titles not extracted yet. I was able to get the SFC games extracted. I also found that Pac-Man Arrangement is based on the Namco Classic Collection Volume 2 ROM set, but it's heavily modified to where MAME will make the game go through a bootloop.

**SFC**
I was looking at #129, and I was thinking: what if the SFC games in this collection also use a similar compression method? To my suprise, they did!

Pac-In-Time and Pac-Attack are stored in their respective DLL plugins as compressed LZMA strings. Using binwalk on both of these will yield working ROMs for both titles, albeit with extra footer data. Both of these games, regardless of if the footers are trimmed, have different CRC32 values than those on the No-Intro database.

Running these binwalk commands in the Plugins folder for the game should get you the extracted titles:

```
#Pac-In-Time
#Outputs 5248D at 1053525 bytes (1029 KiB), CRC32: BC11BF17. Add .sfc file extension to run.

binwalk -o 337037 -Z -S -e "./GameSFC_Pacintime_pacmuseum2021_windows_x64_Release.dll"
	
#Pac-Attack
#Outputs 4CE5D at 524484 bytes (513 KiB), CRC32: 33268046. Add .sfc file extension to run.

binwalk -o 314973 -Z -S -e "./GameSFC_Pacattack_pacmuseum2021_windows_x64_Release.dll"
```


**Pac-Man Arrangement**

Like the SFC games, the Arcade ROMs for Pac-Man Arrangement are stored in a compressed LZMA string that can be extracted with binwalk. binwalk will spit out a file named 67460. 

`binwalk -e "./GamePacmanarrangement_pacmuseum2021_windows_x64_Release.dll"`

Each ROM in this file is contained in the following offsets:

```
ncs2main0.14e - Offset 0x0 - Size 0x80000 - CRC32 A19C715D (BAD CHECKSUM)
ncs2main1.13e - Offset 0x80000 - Size 0x80000 - CRC32 A30BD27C (BAD CHECKSUM)
ncs1sub.1d - Offset 0x100000 - Size 0x80000 - CRC32 365CADBF
ncs1voic.7c - Offset 0x180000 - Size 0x200000 - CRC32 ED05FD88
ncs1cg0.10e - Offset 0x380000 - Size 0x200000 - CRC32 FDD24DBE
ncs1cg1.10f - Offset 1x580000 - Size 0x200000 - CRC32 007B19DE
```

I compressed these into ncv2.zip and ran this in MAME through the command line to get it to ignore the bad checksums. It booted up, but I couldn't get past the boot screen. From what I can tell, sections of the game's data have been removed (mostly Rally-X's and Dig Dug's data), which could be the cause of this.

**Comment by ghost** [comment]
Exciting news about Pac-Man Arrangement! If you play around with the service mode, you can set the game to just Pac-Man Arrangement only and actually play it. 

To do this, enter service mode and go to the following: Game Options > Game Select > Pacman only

Reset the game and it will directly boot up Pac-Man Arrangement!

You can select the other games via the service menu, but unfortunately, as I suspected, all of these games are inaccessible.

**Comment by farmerbb** [comment]
Excellent find! I've verified that Pac-In-Time, Pac-Attack and Pac-Man Arrangement (in MAME) all work using these instructions 👍

Also if anyone is interested, the service menu can also be accessed in the MAME RetroArch core by holding down Y (or X; whichever button's on top) and pressing Select to open the MAME menu.  Then go to DIP Switches -> toggle Service Mode on.  Repeat the steps and toggle Service Mode off when you're done, then choose Reset System in that same menu.

**Comment by DrAzathoth** [comment]
Some misc. info on the SNES ROMs.

- Pac-Attack seems to be based on the US version

- Pac-In-Time is based on the Japanese version and is in Japanese

- Pac-In-Time's intro is partially corrupted due to the collection replacing the graphics for Ms. Pac-Man with Pac-Mom. The footer is abnormally large so it probably includes the new graphics there. Maybe there can be a way to fix the corrupted screens by manually inserting the new graphics ourselves?

**Comment by stuckinstyx** [comment]
Here is a shell script that will help with Pac-Man Arrangement.  It will split up the file created by binwalk, and create a zip that will work with Mame,  It uses forcecrc32.py from Project Nayuki. 

https://www.nayuki.io/page/forcing-a-files-crc-to-any-value

```
#!/bin/bash
mkdir -p temp
dd bs=524288 if=67460 of=temp/ncs2main0.14e skip=0 count=1
dd bs=524288 if=67460 of=temp/ncs2main1.13e skip=1 count=1
dd bs=524288 if=67460 of=temp/ncs1sub.1d skip=2 count=1
dd bs=524288 if=67460 of=temp/ncs1voic.7c skip=3 count=4
dd bs=524288 if=67460 of=temp/ncs1cg0.10e skip=7 count=4
dd bs=524288 if=67460 of=temp/ncs1cg1.10f skip=11 count=4
python3 forcecrc32.py temp/ncs2main0.14e 140 fb8a4123
python3 forcecrc32.py temp/ncs2main1.13e 524284 7a5ef23b
rm -f ncv2.zip
zip ncv2.zip temp/*
```

**Comment by GuiMiTomo** [comment]
YOU GOT THE PAC-MAN ARRANGEMENT ROM?
Oh boy, this is good news.

**Comment by GuiMiTomo** [comment]
If only binwalk worked on Windows... Nonetheless, congrats on your job!

**Comment by DrAzathoth** [comment]
@GuiMiTomo, consider setting up WSL2 on Windows. Very handy for these types of tools and it doesn’t take up much space if you configure it properly!

**Comment by Computerlabrat** [comment]
Cygwin is another option to consider if you want to run Linux programs on Windows.

**Comment by GuiMiTomo** [comment]
Turns out an older version of binwalk worked.
Also, on the results screen of Pac-Attack, it seems that the Ms. Pac-Man and Pac-Baby sprites are glitched. These graphics were also replaced with Pac-Mom and Pac-Sis in PMM+.

**Comment by GuiMiTomo** [comment]
Yep, you read the mention correctly. I did add Arrangement support to my own fork of game-extraction-toolbox.

**Comment by TheNametag** [comment]
It appears that Binwalk had a large change with their version 3, and those command line options do not appear to be there anymore.  Is there a way to get this extraction to work with Binwalk v3?
